### PR TITLE
Fixed message for course-level coupon

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
fixes https://github.com/mitodl/micromasters/issues/3263

#### What's this PR do?
- calculates regular prices for course and show it in coupon message

#### How should this be manually tested?
create coupon and bind it with user, then see the message on dashboard

@pdpinch 
#### Screenshots (if appropriate)
screen shot 2017-06-07 at 5 16 35 pm

### CODE REVIEW

File: static/js/components/dashboard/CourseListCard.js [hunk 1]
```diff
@@ -45,6 +46,26 @@ export default class CourseListCard extends React.Component {
...
+    let list = [];
+
+    for (const courseRun of sortedCourseRuns(program)) {
+      price = prices.get(courseRun.id);
+      if (price) {
+        list.push(price);
+      }
+    }
+    // retruning max price, assuming that max price is regular price otherwise discounted.
+    // what happen when user has coupans for all course in program for COUPON_CONTENT_TYPE_COURSE.
```
Comment (amir-qayyum-khan) [inline static/js/components/dashboard/CourseListCard.js]: @noisecapella

Comment (noisecapella) [inline static/js/components/dashboard/CourseListCard.js]: The way that coupons are structured currently, there should never be more than one coupon per program, even if they are for a course.

File: static/js/components/dashboard/CourseListCard_test.js [hunk 2]
```diff
@@ -144,6 +144,37 @@ describe('CourseListCard', () => {
       );
     });
 
+    it('should show regular course price on coupon message', () => {
+      const coupon = makeCourseCoupon(program["courses"][0].id, program.id);
```
Comment (noisecapella) [inline static/js/components/dashboard/CourseListCard_test.js]: Can you change the syntax here to `program.courses[0].id`?

File: static/js/factories/dashboard.js [hunk 3]
```diff
@@ -129,6 +129,16 @@ export const makeCoupon = (program: Program): Coupon => ({
   object_id: program.id,
 });
 
+export const makeCourseCoupon = (courseId: number, programId: number): Coupon => ({
```
Comment (noisecapella) [inline static/js/factories/dashboard.js]: Can you change `courseId` and `programId` to `course` and `program` for consistency with `makeCoupon`?

**Comment by amir-qayyum-khan** [comment]
> This doesn't fix the problem at all. CourseListCard has a prop prices which has the coupon information already calculated.

@noisecapella  i dont want to change course price there, if discounted price of a course is $50 then let it be.
Here i change the display for coupon message to show regular price instead of first run price

**Comment by noisecapella** [comment]
If you run the `./manage.py snapshot_dashboard_states` command there are some screenshots which illustrate the problem. Here are the results from when I ran the command on your branch: 

[dashboard.tar.gz](https://github.com/mitodl/micromasters/files/1061256/dashboard.tar.gz)

The files with `coupon` in the name relate to various coupon states. I'm changing the filenames in a future PR to be more descriptive but there are two boolean values and the first one is `True` if the coupon is a program coupon and `False` if the coupon is a course coupon. (This is more clear if you look at the source code to the management command.) It shows that the problem still exists in this PR

**Comment by amir-qayyum-khan** [comment]
i am not able to understand it @noisecapella can you guid me what is that problem here in my branch that we want to solve?

**Comment by pdpinch** [comment]
It looks to me like the screenshot `dashboard_state_032_coupon_percent-discount_False_True.png` is incorrect. 

image

It's a little hard to tell for me though because the text file `dashboard_state_032_coupon_percent-discount_False_True.txt` doesn't include the output of the coupons and course_prices APIs. Is that hard to add?

**Comment by noisecapella** [comment]
@pdpinch No, I can look into that

**Comment by amir-qayyum-khan** [comment]
Done @noisecapella 

dashboard_state_026_coupon_fixed-price_course_not-free

dashboard_state_028_coupon_fixed-discount_course_not-free

dashboard_state_030_coupon_percent-discount_course_not-free

dashboard_state_032_coupon_percent-discount_course_free

File: static/js/components/dashboard/CourseListCard.js [hunk 4]
```diff
@@ -47,22 +48,21 @@ export default class CourseListCard extends React.Component {
     checkout:                        (s: string) => void,
   };
 
+  getRegularPrice = (coursePrice?: CoursePrice): Decimal|null => {
+    // assuming all courses have same price
+    return coursePrice ? coursePrice.price : null;
```
Comment (amir-qayyum-khan) [inline static/js/components/dashboard/CourseListCard.js]: https://github.com/mitodl/micromasters/blob/master/static/js/containers/DashboardPage.js#L676

**Comment by noisecapella** [comment]
It looks like the functionality is fixed now, looking at code and tests...

File: static/js/components/dashboard/CourseListCard.js [hunk 5]
```diff
@@ -47,22 +48,20 @@ export default class CourseListCard extends React.Component {
...
 
     if (FA_TERMINAL_STATUSES.includes(finAidStatus)) {
-      let price;
-      for (const courseRun of sortedCourseRuns(program)) {
-        price = prices.get(courseRun.id);
-        if (price) {
-          break;
-        }
-      }
+      let price = this.getRegularPrice(coursePrice);
```
Comment (noisecapella) [inline static/js/components/dashboard/CourseListCard.js]: Can you replace `this.getRegularPrice(coursePrice)` with `R.pathOr(null, ['price'], coursePrice)` here and below?

**Pull request merged**